sns/main.tf

```diff
@@ -0,0 +1,124 @@
+* to let the module create a KMS key it will also include an IAM policy that allows access from the
+* root user of the account. You can also supply service like `s3.amazonaws.com` or `sns.amazonaws.com`
+* in the `kms_event_sources` list, as well as IAM roles in the `kms_iam_sources` list who will then have 
+* `kms:Decrypt*` and `kms:GenerateDataKey*` permissions on the key.
+*/
+
+data "aws_caller_identity" "current" {}
+
+resource "aws_sns_topic" "this" {
+  name                                     = var.name
+  name_prefix                              = var.name_prefix
+  display_name                             = var.display_name
+  policy                                   = var.policy
+  delivery_policy                          = var.delivery_policy
+  application_success_feedback_role_arn    = var.application_success_feedback_role_arn
+  application_success_feedback_sample_rate = var.application_success_feedback_sample_rate
+  application_failure_feedback_role_arn    = var.application_failure_feedback_role_arn
+  http_success_feedback_role_arn           = var.http_success_feedback_role_arn
+  http_success_feedback_sample_rate        = var.http_success_feedback_sample_rate
+  http_failure_feedback_role_arn           = var.http_failure_feedback_role_arn
+  kms_master_key_id                        = var.kms_master_key_id == null ? aws_kms_key.sns_key[0].arn : var.kms_master_key_id
+  fifo_topic                               = var.fifo_topic
+  content_based_deduplication              = var.content_based_deduplication
+  lambda_success_feedback_role_arn         = var.lambda_success_feedback_role_arn
+  lambda_success_feedback_sample_rate      = var.lambda_success_feedback_sample_rate
+  lambda_failure_feedback_role_arn         = var.lambda_failure_feedback_role_arn
+  sqs_success_feedback_role_arn            = var.sqs_success_feedback_role_arn
+  sqs_success_feedback_sample_rate         = var.sqs_success_feedback_sample_rate
+  sqs_failure_feedback_role_arn            = var.sqs_failure_feedback_role_arn
+  firehose_success_feedback_role_arn       = var.firehose_success_feedback_role_arn
+  firehose_success_feedback_sample_rate    = var.firehose_success_feedback_sample_rate
+  firehose_failure_feedback_role_arn       = var.firehose_failure_feedback_role_arn
+  tags                                     = merge(local.common_tags, var.tags)
+}
+
+
+
+data "aws_iam_policy_document" "kms_policies" {
+
+  statement {
+
+    effect = "Allow"
+
+    actions = [
+      "kms:*"
+    ]
+
+    resources = [
+      "*"
+    ]
+
+    principals {
+      type        = "AWS"
+      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"]
+    }
+  }
+
+  dynamic "statement" {
+    for_each = var.kms_event_sources
+
+    content {
+      effect = "Allow"
+
+      principals {
+        type        = "Service"
+        identifiers = [statement.value]
+      }
```

Comment:
Is it required to do a for each on this or can we simply do something like 

```hcl 
      principals {
        type        = "Service"
        identifiers = [var.kms_event_sources]
      }
```

and have the count be something like:

```hcl
   count = (length(var.kms_event_sources) > 0) ? 1 : 0
```

There is a limit to the number characters a policy can have.

It is 6144 characters so we have room, so not a big deal if we can't.

https://aws.amazon.com/premiumsupport/knowledge-center/iam-increase-policy-size/#:~:text=For%20more%20information%2C%20see%20IAM,for%20a%20service%20quota%20increase.